apps/ejabberd/src/mod_mam_utils.erl

```diff
@@ -668,6 +680,34 @@ form_field(Type, VarName) ->
                     {<<"var">>, VarName}]}.
 
 %% -----------------------------------------------------------------------
+%% Text search tokenization
+
+%% -----------------------------------------------------------------------
+%% @doc
+%% Normalize given text to improve text search in some MAM backends.
+%% This normalization involves making text all lowercase, replacing some word separators
+%% ([, .:;-?!]) with given one (by default "%") and removing all unicode characters that are
+%% considered non-alphanumerical.
+%% For example, text: "My cat, was eaten by: my dog?!? Why...?!?" will be normalized as:
+%% "my%cat%was%eaten%by%my%dog%why"
+%% @end
+%% -----------------------------------------------------------------------
+-spec normalize_search_text(binary() | string() | undefined) -> string() | undefined.
+normalize_search_text(Text) ->
+    normalize_search_text(Text, "%").
+
+-spec normalize_search_text(binary() | string() | undefined, string()) -> string() | undefined.
+normalize_search_text(undefined, _WordSeparator) ->
+    undefined;
+normalize_search_text(Text, WordSeparator) ->
```

Comment:
It adds like 100 μs to processing time per row on my slow CPU. Simple `string:to_lower` is only 10 times faster then this whole function. Also, I've checked that precompiling regexes only speed up this function by ~5%. Removing only one of those 3 regexes could give also no more then 5-10%, but I'm not sure if its easy to do.